File: packages/runtime-kubernetes/src/translator.ts
import type { Workload } from "@boilerhouse/core";
import type { InstanceId } from "@boilerhouse/core";
import type { K8sPod, K8sService, K8sContainer, K8sVolume, K8sProbe } from "./types";

export const MANAGED_LABEL = "boilerhouse.dev/managed";
export const INSTANCE_ID_LABEL = "boilerhouse.dev/instance-id";
export const WORKLOAD_NAME_LABEL = "boilerhouse.dev/workload-name";

export interface TranslationResult {
	pod: K8sPod;
	service?: K8sService;
}

/**
 * Translates a Boilerhouse {@link Workload} into K8s Pod (and optional Service) specs.
 *
 * Mapping:
 * - `image.ref` → container image
 * - `resources` → resource limits
 * - `network.expose` → container ports + Service
 * - `entrypoint` → command/args/env
 * - `filesystem.overlay_dirs` → emptyDir volumes
 * - `health` → readinessProbe
 */
/**
 * @param imageOverride Resolved image ref, used when the workload was built from a Dockerfile.
 */
export function workloadToPod(
	workload: Workload,
	instanceId: InstanceId,
	namespace: string,
	imageOverride?: string,
): TranslationResult {
	const labels: Record<string, string> = {
		[MANAGED_LABEL]: "true",
		[INSTANCE_ID_LABEL]: instanceId,
		[WORKLOAD_NAME_LABEL]: workload.workload.name,
	};

	const imageRef = imageOverride ?? workload.image.ref;
	if (!imageRef) {
		throw new Error(
			`Workload "${workload.workload.name}" has no image ref and no imageOverride was provided`,
		);
	}

	const container: K8sContainer = {
		name: "main",
		image: imageRef,
	};

	// Resources
	container.resources = {
		limits: {
			cpu: `${workload.resources.vcpus * 1000}m`,
			memory: `${workload.resources.memory_mb}Mi`,
		},
		requests: {
			cpu: `${Math.max(100, Math.floor(workload.resources.vcpus * 250))}m`,
			memory: `${Math.min(workload.resources.memory_mb, 128)}Mi`,
		},
	};

	// Entrypoint
	if (workload.entrypoint) {
		if (workload.entrypoint.cmd) {
			container.command = [workload.entrypoint.cmd];
		}
		if (workload.entrypoint.args) {
			container.args = workload.entrypoint.args;
		}
		if (workload.entrypoint.env) {
			container.env = Object.entries(workload.entrypoint.env).map(
				([name, value]) => ({ name, value }),
			);
		}
		if (workload.entrypoint.workdir) {
			container.workingDir = workload.entrypoint.workdir;
		}
	}

	// Ports
	const exposePorts = workload.network?.expose;
	if (exposePorts && exposePorts.length > 0) {
		container.ports = exposePorts.map((p) => ({
			containerPort: p.guest,
			protocol: "TCP",
		}));
	}

	// Health check → readinessProbe
	if (workload.health) {
		const probe: K8sProbe = {
			periodSeconds: workload.health.interval_seconds,
			failureThreshold: workload.health.unhealthy_threshold,
		};
		if (workload.health.check_timeout_seconds) {
			probe.timeoutSeconds = workload.health.check_timeout_seconds;
		}

		if (workload.health.http_get) {
			probe.httpGet = {
				path: workload.health.http_get.path,
				port: workload.health.http_get.port ?? (exposePorts?.[0]?.guest ?? 8080),
			};
		} else if (workload.health.exec) {
			probe.exec = { command: workload.health.exec.command };
		}

		container.readinessProbe = probe;
	}

	// Volumes from overlay_dirs
	const volumes: K8sVolume[] = [];
	const volumeMounts: Array<{ name: string; mountPath: string }> = [];

	if (workload.filesystem?.overlay_dirs) {
		for (let i = 0; i < workload.filesystem.overlay_dirs.length; i++) {
			const dir = workload.filesystem.overlay_dirs[i]!;
			const name = `overlay-${i}`;
			volumes.push({ name, emptyDir: { sizeLimit: "256Mi" } });
			volumeMounts.push({ name, mountPath: dir });
		}
	}

	if (volumeMounts.length > 0) {
		container.volumeMounts = volumeMounts;
	}

	const pod: K8sPod = {
		apiVersion: "v1",
		kind: "Pod",
		metadata: {
			name: instanceId,
			namespace,
			labels,
		},
		spec: {
			containers: [container],
			restartPolicy: "Never",
			...(volumes.length > 0 ? { volumes } : {}),
		},
	};

	// Service for exposed ports.
	// Service names must be valid DNS-1035 labels (start with a letter),
	// but instance IDs are UUIDs that may start with a digit. Prefix with "svc-".
	let service: K8sService | undefined;
	if (exposePorts && exposePorts.length > 0) {
		service = {
			apiVersion: "v1",
			kind: "Service",
			metadata: {
				name: `svc-${instanceId}`,
				namespace,
				labels,
			},
			spec: {
				selector: { [INSTANCE_ID_LABEL]: instanceId },
				ports: exposePorts.map((p, i) => ({
					port: p.guest,
					targetPort: p.guest,
					protocol: "TCP",
					...(exposePorts.length > 1 ? { name: `port-${i}` } : {}),
				})),
				type: "ClusterIP",
			},
		};
	}

	return { pod, service };
}

File: go/internal/operator/translator.go
package operator

import (
	"encoding/json"
	"fmt"
	"sort"

	"github.com/zdavison/boilerhouse/go/internal/envoy"

	v1alpha1 "github.com/zdavison/boilerhouse/go/api/v1alpha1"
	corev1 "k8s.io/api/core/v1"
	networkingv1 "k8s.io/api/networking/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/util/intstr"
)

const (
	LabelManaged    = "boilerhouse.dev/managed"
	LabelWorkload   = "boilerhouse.dev/workload"
	LabelInstance   = "boilerhouse.dev/instance"
	LabelTenant     = "boilerhouse.dev/tenant"
	LabelPoolStatus = "boilerhouse.dev/pool-status"

	linkLocalCIDR = "169.254.0.0/16"
)

// TranslateOpts holds the metadata needed to translate a workload spec into
// Kubernetes resources.
type TranslateOpts struct {
	InstanceId   string
	WorkloadName string
	TenantId     string // empty for pool pods
	Namespace    string
	PoolStatus   string // "warming", "ready", or "" for non-pool
	ProxyConfig  *ProxyConfig // nil if no sidecar needed
}

// TranslateResult holds the Kubernetes resources produced by Translate.
type TranslateResult struct {
	Pod           *corev1.Pod
	Service       *corev1.Service
	NetworkPolicy *networkingv1.NetworkPolicy
	ConfigMap     *corev1.ConfigMap // for Envoy sidecar proxy config
}

// Translate converts a BoilerhouseWorkloadSpec and metadata into Kubernetes
// resource specs: Pod, optional Service, NetworkPolicy, and optional PVC.
func Translate(spec v1alpha1.BoilerhouseWorkloadSpec, opts TranslateOpts) (*TranslateResult, error) {
	result := &TranslateResult{}

	labels := buildLabels(opts)

	pod, err := buildPod(spec, opts, labels)
	if err != nil {
		return nil, fmt.Errorf("building pod: %w", err)
	}
	result.Pod = pod

	if spec.Network != nil && len(spec.Network.Expose) > 0 {
		result.Service = buildService(spec, opts, labels)
	}

	result.NetworkPolicy = buildNetworkPolicy(spec, opts, labels)

	// PVC creation removed: overlay persistence now uses snapshot-based approach.
	// Overlay dirs use emptyDir volumes in the Pod spec (see buildPod).

	// Inject Envoy sidecar if ProxyConfig is provided.
	if opts.ProxyConfig != nil {
		configMapName := fmt.Sprintf("proxy-%s", opts.InstanceId)
		InjectSidecar(result.Pod, configMapName)
		result.ConfigMap = buildProxyConfigMap(opts, labels, configMapName)
	}

	return result, nil
}

func buildLabels(opts TranslateOpts) map[string]string {
	labels := map[string]string{
		LabelManaged:  "true",
		LabelWorkload: opts.WorkloadName,
		LabelInstance: opts.InstanceId,
	}
	if opts.TenantId != "" {
		labels[LabelTenant] = opts.TenantId
	}
	if opts.PoolStatus != "" {
		labels[LabelPoolStatus] = opts.PoolStatus
	}
	return labels
}

func buildPod(spec v1alpha1.BoilerhouseWorkloadSpec, opts TranslateOpts, labels map[string]string) (*corev1.Pod, error) {
	falseVal := false
	terminationGrace := int64(3)

	container, err := buildContainer(spec, opts)
	if err != nil {
		return nil, err
	}

	pod := &corev1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      opts.InstanceId,
			Namespace: opts.Namespace,
			Labels:    labels,
		},
		Spec: corev1.PodSpec{
			Containers:                    []corev1.Container{*container},
			RestartPolicy:                 corev1.RestartPolicyNever,
			AutomountServiceAccountToken:  &falseVal,
			TerminationGracePeriodSeconds: &terminationGrace,
			SecurityContext: &corev1.PodSecurityContext{
				SeccompProfile: &corev1.SeccompProfile{
					Type: corev1.SeccompProfileTypeRuntimeDefault,
				},
			},
		},
	}

	// Overlay volume mounts: use emptyDir volumes for all overlay dirs.
	// Persistence is handled by the snapshot manager (tar extract/inject),
	// not per-tenant PVCs.
	if spec.Filesystem != nil && len(spec.Filesystem.OverlayDirs) > 0 {
		var volumes []corev1.Volume
		var mounts []corev1.VolumeMount
		for i, dir := range spec.Filesystem.OverlayDirs {
			volName := fmt.Sprintf("overlay-%d", i)
			volumes = append(volumes, corev1.Volume{
				Name: volName,
				VolumeSource: corev1.VolumeSource{
					EmptyDir: &corev1.EmptyDirVolumeSource{},
				},
			})
			mounts = append(mounts, corev1.VolumeMount{
				Name:      volName,
				MountPath: dir,
			})
		}
		pod.Spec.Volumes = append(pod.Spec.Volumes, volumes...)
		pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, mounts...)
	}

	return pod, nil
}

func buildContainer(spec v1alpha1.BoilerhouseWorkloadSpec, opts TranslateOpts) (*corev1.Container, error) {
	falseVal := false

	container := &corev1.Container{
		Name:  "main",
		Image: spec.Image.Ref,
		SecurityContext: &corev1.SecurityContext{
			Capabilities: &corev1.Capabilities{
				Drop: []corev1.Capability{"ALL"},
			},
			AllowPrivilegeEscalation: &falseVal,
		},
	}

	// Resources
	cpuMillis := spec.Resources.VCPUs * 1000
	memMi := spec.Resources.MemoryMb

	cpuReqMillis := spec.Resources.VCPUs * 250
	if cpuReqMillis < 100 {
		cpuReqMillis = 100
	}
	memReqMi := memMi
	if memReqMi > 128 {
		memReqMi = 128
	}

	container.Resources = corev1.ResourceRequirements{
		Limits: corev1.ResourceList{
			corev1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%dm", cpuMillis)),
			corev1.ResourceMemory: resource.MustParse(fmt.Sprintf("%dMi", memMi)),
		},
		Requests: corev1.ResourceList{
			corev1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%dm", cpuReqMillis)),
			corev1.ResourceMemory: resource.MustParse(fmt.Sprintf("%dMi", memReqMi)),
		},
	}

	// Entrypoint
	if spec.Entrypoint != nil {
		if spec.Entrypoint.Cmd != "" {
			container.Command = []string{spec.Entrypoint.Cmd}
		}
		if len(spec.Entrypoint.Args) > 0 {
			container.Args = spec.Entrypoint.Args
		}
		if spec.Entrypoint.Workdir != "" {
			container.WorkingDir = spec.Entrypoint.Workdir
		}
		if spec.Entrypoint.Env != nil && spec.Entrypoint.Env.Raw != nil {
			envVars, err := parseEnvVars(spec.Entrypoint.Env.Raw)
			if err != nil {
				return nil, fmt.Errorf("parsing env vars: %w", err)
			}
			container.Env = envVars
		}
	}

	// Exposed ports
	if spec.Network != nil {
		for _, p := range spec.Network.Expose {
			container.Ports = append(container.Ports, corev1.ContainerPort{
				ContainerPort: int32(p.Guest),
				Protocol:      corev1.ProtocolTCP,
			})
		}
	}

	// Health probe
	if spec.Health != nil {
		probe := &corev1.Probe{
			PeriodSeconds:  int32(spec.Health.IntervalSeconds),
			FailureThreshold: int32(spec.Health.UnhealthyThreshold),
		}
		if spec.Health.HTTPGet != nil {
			probe.ProbeHandler = corev1.ProbeHandler{
				HTTPGet: &corev1.HTTPGetAction{
					Path: spec.Health.HTTPGet.Path,
					Port: intstr.FromInt32(int32(spec.Health.HTTPGet.Port)),
				},
			}
		} else if spec.Health.Exec != nil {
			probe.ProbeHandler = corev1.ProbeHandler{
				Exec: &corev1.ExecAction{
					Command: spec.Health.Exec.Command,
				},
			}
		}
		container.ReadinessProbe = probe
	}

	return container, nil
}

// parseEnvVars decodes a JSON object of env vars (map[string]string) into
// a sorted slice of corev1.EnvVar.
func parseEnvVars(raw []byte) ([]corev1.EnvVar, error) {
	var m map[string]string
	if err := json.Unmarshal(raw, &m); err != nil {
		return nil, err
	}

	// Sort keys for deterministic output.
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	envVars := make([]corev1.EnvVar, 0, len(m))
	for _, k := range keys {
		envVars = append(envVars, corev1.EnvVar{
			Name:  k,
			Value: m[k],
		})
	}
	return envVars, nil
}

func buildService(spec v1alpha1.BoilerhouseWorkloadSpec, opts TranslateOpts, labels map[string]string) *corev1.Service {
	var ports []corev1.ServicePort
	for i, p := range spec.Network.Expose {
		sp := corev1.ServicePort{
			Port:       int32(p.Guest),
			TargetPort: intstr.FromInt32(int32(p.Guest)),
			Protocol:   corev1.ProtocolTCP,
		}
		if len(spec.Network.Expose) > 1 {
			sp.Name = fmt.Sprintf("port-%d", i)
		}
		ports = append(ports, sp)
	}

	return &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      opts.InstanceId,
			Namespace: opts.Namespace,
			Labels:    labels,
		},
		Spec: corev1.ServiceSpec{
			Selector: map[string]string{
				LabelInstance: opts.InstanceId,
			},
			Ports: ports,
			Type:  corev1.ServiceTypeClusterIP,
		},
	}
}

func buildNetworkPolicy(spec v1alpha1.BoilerhouseWorkloadSpec, opts TranslateOpts, labels map[string]string) *networkingv1.NetworkPolicy {
	access := "none"
	if spec.Network != nil && spec.Network.Access != "" {
		access = spec.Network.Access
	}

	np := &networkingv1.NetworkPolicy{
		ObjectMeta: metav1.ObjectMeta{
			Name:      opts.InstanceId,
			Namespace: opts.Namespace,
			Labels:    labels,
		},
		Spec: networkingv1.NetworkPolicySpec{
			PodSelector: metav1.LabelSelector{
				MatchLabels: map[string]string{
					LabelInstance: opts.InstanceId,
				},
			},
			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeEgress},
		},
	}

	switch access {
	case "unrestricted":
		np.Spec.Egress = []networkingv1.NetworkPolicyEgressRule{
			dnsEgressRule(),
			{
				To: []networkingv1.NetworkPolicyPeer{
					{
						IPBlock: &networkingv1.IPBlock{
							CIDR:   "0.0.0.0/0",
							Except: []string{linkLocalCIDR},
						},
					},
				},
			},
		}
	case "restricted":
		np.Spec.Egress = []networkingv1.NetworkPolicyEgressRule{
			dnsEgressRule(),
			{
				Ports: []networkingv1.NetworkPolicyPort{
					networkPolicyPort(corev1.ProtocolTCP, 443),
				},
				To: []networkingv1.NetworkPolicyPeer{
					{
						IPBlock: &networkingv1.IPBlock{
							CIDR:   "0.0.0.0/0",
							Except: []string{linkLocalCIDR},
						},
					},
				},
			},
		}
	default: // "none" — deny all egress
		np.Spec.Egress = []networkingv1.NetworkPolicyEgressRule{}
	}

	return np
}

func dnsEgressRule() networkingv1.NetworkPolicyEgressRule {
	return networkingv1.NetworkPolicyEgressRule{
		Ports: []networkingv1.NetworkPolicyPort{
			networkPolicyPort(corev1.ProtocolUDP, 53),
			networkPolicyPort(corev1.ProtocolTCP, 53),
		},
		To: []networkingv1.NetworkPolicyPeer{
			{NamespaceSelector: &metav1.LabelSelector{}},
		},
	}
}

func networkPolicyPort(protocol corev1.Protocol, port int) networkingv1.NetworkPolicyPort {
	p := intstr.FromInt32(int32(port))
	proto := protocol
	return networkingv1.NetworkPolicyPort{
		Protocol: &proto,
		Port:     &p,
	}
}


// buildProxyConfigMap creates a ConfigMap with the Envoy YAML config and CA cert
// for the sidecar proxy.
func buildProxyConfigMap(opts TranslateOpts, labels map[string]string, name string) *corev1.ConfigMap {
	data := map[string]string{
		"envoy.yaml": opts.ProxyConfig.EnvoyYAML,
	}
	if opts.ProxyConfig.CACert != nil {
		data["ca.crt"] = string(opts.ProxyConfig.CACert)
	}

	// Add per-domain TLS certs if TLS material is available.
	if opts.ProxyConfig.TLS != nil {
		for _, dc := range opts.ProxyConfig.TLS.Certs {
			safe := envoy.SafeDomain(dc.Domain)
			data["certs/"+safe+".crt"] = string(dc.Cert)
			data["certs/"+safe+".key"] = string(dc.Key)
		}
	}

	return &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: opts.Namespace,
			Labels:    labels,
		},
		Data: data,
	}
}
